### PR TITLE
Modify std.conv.octal to be usable at runtime

### PR DESCRIPTION
**Pull request by McSherry**
Currently, `std.conv.octal` is only usable with compile-time parameters. This change should be fairly minor, and makes `octal` usable at runtime.

### CODE REVIEW

File: std/conv.d [hunk 1]
```diff
@@ -3737,21 +3737,44 @@ unittest
...
+    if (!isOctalLiteralString(num) &&
+        // Should help alleviate what are likely to be reasonably common
+        // mistakes. For example, feeding the result of 'readln' directly
+        // in to 'octal' would cause a failure if this wasn't here because
+        // it would include the newline character. Happy to remove if this
+        // is determined to be outside the function's remit.
+        !isOctalLiteralString((num = num.strip))
+        )
+    {
+        throw new ConvException(
```
Comment (MetaLang) [inline std/conv.d]: Please use an assert here instead so `octal` can be used in `nothrow` code.

**Comment by schveiguy** [comment]
somehow there are some trivial white space changes, can you remove those?

File: std/conv.d [hunk 2]
```diff
@@ -3737,21 +3737,44 @@ unittest
 @property T octal(T, string num)()
     if (isOctalLiteral!num)
 {
+    return octal!T(num);
+}
+
+/// Ditto
+@property T octal(T)(string num)
```
Comment (JakobOvrum) [inline std/conv.d]: Use `in char[]` (or `const(char)[]`) as immutability is not required.

**Comment by McSherry** [comment]
Changes made.

Trivial white-space was probably the editor I was using trying to be helpful.

As an aside, for some reason the `Darwin_64_64` and `FreeBSD_64` builds seems to be assuming that the integer literal on [line 3754](https://github.com/McSherry/phobos/blob/runtime-octal-conv/std/conv.d#L3754) is a `ulong`.

**Comment by schveiguy** [comment]
Right. It's because it is not known at compile time any more, and num.length is size_t (ulong on 64 bit systems)

Just change type in loop from int to auto

**Comment by schveiguy** [comment]
Sorry, not auto but ptrdiff_t!

**Comment by McSherry** [comment]
Made the adjustment.

**Comment by MetaLang** [comment]
LGTM

File: std/conv.d [hunk 3]
```diff
@@ -3737,10 +3737,21 @@ unittest
 @property T octal(T, string num)()
     if (isOctalLiteral!num)
 {
+    return octal!T(num);
+}
+
+/// Ditto
+@property T octal(T)(string num) pure nothrow
```
Comment (JakobOvrum) [inline std/conv.d]: This still needs to be a `const`, not `immutable`, parameter type.

**Comment by McSherry** [comment]
> This still needs to be a `const`, not `immutable`, parameter type.

Sorry, must have slipped my mind. Changed the type.

**Comment by schveiguy** [comment]
Looks good, can you squash this into a few commits?

**Comment by McSherry** [comment]
Squashed.

File: std/conv.d [hunk 4]
```diff
@@ -3737,10 +3737,21 @@ unittest
...
     if (isOctalLiteral!num)
 {
+    return octal!T(num);
+}
+
+/// Ditto
+@property T octal(T)(const(char)[] num) pure nothrow
+    if (isNumeric!T)
+{
+    assert( isOctalLiteralString(num),
```
Comment (JakobOvrum) [inline std/conv.d]: This precondition must be documented. The result will be undefined when `num` contains non-octal-digit-characters.

Comment (McSherry) [inline std/conv.d]: What is the preferred way to document assertions?

File: std/conv.d [hunk 5]
```diff
@@ -3737,10 +3737,21 @@ unittest
 @property T octal(T, string num)()
     if (isOctalLiteral!num)
 {
+    return octal!T(num);
+}
+
+/// Ditto
+@property T octal(T)(const(char)[] num) pure nothrow
```
Comment (andralex) [inline std/conv.d]: should accept any range of characters

File: std/conv.d [hunk 6]
```diff
@@ -3812,7 +3825,7 @@ Returns if the given string is a correctly formatted octal literal.
 The format is specified in lex.html. The leading zero is allowed, but
 not required.
  */
-bool isOctalLiteralString(string num)
+bool isOctalLiteralString(const(char)[] num) pure nothrow
```
Comment (andralex) [inline std/conv.d]: ditto

File: std/conv.d (hunk 5 above)
Comment (McSherry) [inline std/conv.d]: Do you mean range as in `InputRange`/`ForwardRange` or as in `char`/`wchar`/`dchar`?

File: std/conv.d (hunk 4 above)
Comment (JakobOvrum) [inline std/conv.d]: A `Preconditions:` DDoc block is probably a good way to place such information.

File: std/conv.d (hunk 5 above)
Comment (schveiguy) [inline std/conv.d]: Can we make this a separate pull? I don't think all the kinks of using generic char ranges are worked out yet, or am I wrong?

**Comment by schveiguy** [comment]
I think it's OK to merge this, we can worry about rangifying octal later. There are many different overloads for octal that need to be considered, it's not trivial.

**Comment by schveiguy** [comment]
Auto-merge toggled on

**Pull request merged**